org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/rename/LSPInlineRenameLinkedMode.java

```diff
@@ -0,0 +1,460 @@
+									document.replace(pos.getOffset(), pos.getLength(), original);
+								}
+							} catch (final BadLocationException ex) {
+								LanguageServerPlugin.logError(ex);
+							} finally {
+								if (widget != null && !widget.isDisposed()) {
+									widget.setRedraw(true);
+								}
+							}
+						});
+					} else {
+						scheduleRenameJob();
+					}
+				}
+
+				@Override
+				public void resume(LinkedModeModel environment, int flags) {
+					// no-op
+				}
+
+				@Override
+				public void suspend(LinkedModeModel environment) {
+					// no-op
+				}
+			});
+			model.forceInstall();
+
+			final var ui = new EditorLinkedModeUI(model, viewer);
+			ui.setExitPolicy(this::exitPolicy);
+			ui.enter();
+		} catch (final BadLocationException ex) {
+			LanguageServerPlugin.logError(ex);
+		}
+	}
+
+	private void scheduleRenameJob() {
+		final var job = new Job(Messages.rename_title) {
+			@Override
+			protected IStatus run(final IProgressMonitor monitor) {
+				try {
+					// LinkedPosition tracks the live identifier region while the user is typing.
+					// We must use it (not the original Region) because the length/offset can
+					// change.
+					final var position = linkedPosition;
+					if (position == null) {
+						LanguageServerPlugin.logWarning("Inline rename: linkedPosition is null, skipping"); //$NON-NLS-1$
+						return Status.OK_STATUS;
+					}
+
+					final String newName = document.get(position.getOffset(), position.getLength());
+					if (newName.equals(originalName)) {
+						return Status.OK_STATUS;
+					}
+
+					/*
+					 * Perform revert + LSP rename as a single visual step on the UI thread.
+					 *
+					 * Language servers expect the "old" identifier to still be present at the
+					 * rename position when textDocument/rename is invoked. If we call rename while
+					 * the buffer already contains the new name, language servers will either treat
+					 * it as a no-op or fail to resolve the symbol, and references will not be
+					 * updated.
+					 *
+					 * To avoid that, we:
+					 *
+					 * 1) Restore the original identifier in the document so the LS sees the
+					 * pre-rename state.
+					 *
+					 * 2) Immediately call textDocument/rename(newName) and apply the resulting
+					 * WorkspaceEdit, which updates all occurrences (including this one).
+					 *
+					 * To prevent a visible flicker in the active editor, we wrap the revert + apply
+					 * sequence in a redraw block for this editor's StyledText: setRedraw(false) ...
+					 * setRedraw(true).
+					 */
+					UI.runOnUIThread(() -> {
+						final StyledText widget = viewer.getTextWidget();
+						if (widget == null || widget.isDisposed()) {
+							boolean success = false;
+							try {
+								document.replace(position.getOffset(), position.getLength(), originalName);
+								success = doRename(newName);
+							} catch (BadLocationException e) {
+								LanguageServerPlugin.logError(e);
+							} finally {
+								if (!success) {
```

Comment:
I am a bit surprised by this. Since `doRename` catches already the expected exceptions, why do we need the success variable out of the try block so that we can use it in `finally`? Can we not the check just inside the try block, same as in lines 428-431?